src/plugins/kibana_utils/public/url/index.ts

```diff
@@ -140,30 +141,52 @@ export function setStateToUrl<T extends BaseState>(
  * listen(cb) - accepts a callback which will be called whenever url has changed
  * update(url: string, replace: boolean) - get an absolute / relative url to update the location to
  */
-export const createUrlControls = () => {
+interface IUrlControls {
+  listen: (cb: () => void) => () => void;
+  update: (updater: (currentUrl: string) => string, replace: boolean) => Promise<string>;
+}
+
+let urlControls: IUrlControls;
+export const getUrlControls = () => {
+  if (urlControls) return urlControls;
+
   const history = createBrowserHistory();
-  return {
+  const updateQueue: Array<(currentUrl: string) => string> = [];
+  return (urlControls = {
     listen: (cb: () => void) =>
       history.listen(() => {
         cb();
       }),
-    update: (url: string, replace = false) => {
-      const { pathname, search } = parseUrl(url);
-      const parsedHash = parseUrlHash(url);
-      const searchQueryString = stringifyQueryString(parsedHash.query);
-      const location = {
-        pathname,
-        hash: formatUrl({
-          pathname: parsedHash.pathname,
-          search: searchQueryString,
-        }),
-        search,
-      };
-      if (replace) {
-        history.replace(location);
-      } else {
-        history.push(location);
-      }
+    update: (updater: (currentUrl: string) => string, replace = false) => {
+      updateQueue.push(updater);
+
+      return Promise.resolve().then(() => {
```

Comment:
This is the main idea. 

As issue with batched writing to storage seems URL specific (because of history) - then it makes sense to solve it on URL level. 
`update()` - becomes async. And it resolves when it actually updates url. It delays url update to the next micro task allowing all synchronous code to compete.